### DISCUSSION TITLE
Image.convert broken for P to RGB?

### DISCUSSION THREAD

**Discussion by TheRealQuantam** [python-pillow/Pillow#6390] [Q&A] [answered]
I'm encountering several severe issues with palettized color in Pillow 9.1.1 on Python 3.7 64-bit. I'm posting this in discussions in case I am doing something wrong, but if I'm not, I'll elevate these to issues. Per request I'm splitting off each issue into a separate thread.

Image.convert from P to RGB behaves like L to RGB and DOES NOT use Image.palette.

Partial code:
```python
img = Image.fromarray(bitmap, "P")
img.palette = img_pal
img.convert("RGB").save("converted as greyscale not palette.png")

```
[bad_pillow.zip](https://github.com/python-pillow/Pillow/files/8954031/bad_pillow.zip)

Correct version:
![manually remapped palette](https://user-images.githubusercontent.com/54358053/174932980-ee0b10d6-ce1a-431d-90f0-08e4e41c0efc.gif)

Converted as L to RGB when it should have been converted as P to RGB:
![converted as greyscale not palette](https://user-images.githubusercontent.com/54358053/174932979-9e24a988-e0e8-42a0-bd5f-c1609a5fabb5.png)

**Comment by radarhere** [comment, marked as answer]
Testing your code, I think all of your problems can be solved by replacing
```python
img.palette = img_pal
```
with
```python
img.putpalette(img_pal)
```
See https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.putpalette for documentation.

**Reply by TheRealQuantam** [reply]
Confirmed that this issue is fixed by putpalette. I'll update the other two with the problems remaining.

Thanks for the help.

**Reply by TheRealQuantam** [reply]
Might it be a good idea to make Image.convert fail if the image is set to P but no palette has been properly set? Silently treating it as L is really confusing.

**Reply by radarhere** [reply]
If you run
```python
img = Image.fromarray(bitmap, "P")
print(img.palette.palette)
```
you will see that there is a palette, it's just empty. It doesn't sound right to say that an empty palette is invalid.

Also, the next release of Pillow will change the behaviour of `fromarray` with "P" mode so that the palette is populated by default. See #6283.

**Reply by TheRealQuantam** [reply]
I see. Okay. Thanks for the help.
